integration-tests/chaos/congestion/simulator.go

```diff
@@ -0,0 +1,484 @@
+	})
+	return nil
+}
+
+func (s *Simulator) startTransactionProducer(ctx context.Context, wg *sync.WaitGroup) {
+	numberOfSenders := s.NumberOfTxsPerBlock
+	work := make(chan struct{}, numberOfSenders)
+	wg.Add(numberOfSenders)
+	for i := range numberOfSenders {
+		go s.runSender(ctx, wg, i, work)
+	}
+
+	txsToSend := numberOfSenders / 2
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		s.listenHeadsUntilCanceled(ctx, func(ctx context.Context, head *types.Header) error {
+			congestion := float64(head.GasUsed) / float64(head.GasLimit)
+			congestionTarget := s.expectedCurrentCongestion.Load().(float64)
+			if txsToSend < numberOfSenders && congestion < congestionTarget*0.95 {
+				txsToSend++
+			} else if txsToSend > 0 && congestion > congestionTarget*1.05 {
+				txsToSend--
+			}
+
+			s.lggr.Debugf("Sending %d transactions; current congestion: %.2f target: %.2f", txsToSend, congestion, congestionTarget)
+			for range txsToSend {
+				select {
+				case work <- struct{}{}:
+				case <-ctx.Done():
+					return nil
+				}
+			}
+			return nil
+		})
+	}()
+}
+
+func (s *Simulator) runPhaseController(ctx context.Context, wg *sync.WaitGroup) {
+	defer wg.Done()
+	s.lggr.Info("Starting ramp up phase")
+	s.phase.Store(phaseRampUp)
+	s.updateCurrentFees(0)
+	s.handleVolatilePhase(ctx, s.RampUp, true)
+
+	s.lggr.Info("Starting plateau phase")
+	s.phase.Store(phasePlateau)
+	s.expectedCurrentCongestion.Store(s.Plateau.Congestion)
+	timer := time.NewTimer(time.Second * time.Duration(s.Plateau.Duration))
+	defer timer.Stop()
+	select {
+	case <-timer.C:
+		break
+	case <-ctx.Done():
+		return
+	}
+
+	s.lggr.Info("Starting cooldown phase")
+	s.phase.Store(phaseCoolDown)
+	s.handleVolatilePhase(ctx, s.CoolDown, false)
+
+	s.lggr.Info("Finished iteration")
+	s.phase.Store(phaseInactive)
+}
+
+func (s *Simulator) handleVolatilePhase(ctx context.Context, phase Phase, isIncrease bool) {
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+	phaseTimer := time.NewTimer(time.Second * time.Duration(phase.Duration))
+	defer phaseTimer.Stop()
+	phaseStart := time.Now()
+	s.expectedCurrentCongestion.Store(phase.Congestion)
+phaseLoop:
+	for {
+		select {
+		case <-ticker.C:
+		case <-phaseTimer.C:
+			break phaseLoop
+		case <-ctx.Done():
+			return
+		}
+
+		elapsedRatio := min(float64(time.Since(phaseStart))/float64(time.Duration(phase.Duration)*time.Second), 1)
+		if !isIncrease {
+			// count backwards. If 0.1 of cooldown phase has passed, we should be at the same price point as if 0.9 of ramp up phase has passed
+			elapsedRatio = 1 - elapsedRatio
+		}
+		s.updateCurrentFees(elapsedRatio)
+	}
+
+	if isIncrease {
+		s.updateCurrentFees(1)
+	} else {
+		s.updateCurrentFees(0)
+	}
+}
+
+// updateCurrentFees - sets fees according to expected distance from target. If targetDistanceRatio = 1, current fees s.FeesIncreasePercent compared to initial.
+func (s *Simulator) updateCurrentFees(targetDistanceRatio float64) {
```

Comment:
wdyt about declaring two constants like this? and then use that instead of `0` and `1`?
```
const (
     IncreasingRatio = 1.0
     FlatRatio = 0.0
)
```

(feel free to use other names).

Also, am I reading it correctly that if it's cooldown phase we are not increasing fees anymore? Shouldn't we be decreasing them? Or we expect them to fall only due to sending less transactions?

Reply:
Hmm, I'm not sure about the constants, but I agree that `updateCurrentFees` and `handleVolatilePhase` are confusing. I've refactored them to be more readable.